### ISSUE TITLE
[BUG] JsonConverterAttribute is ignored on polymorphic base class properties

### ISSUE THREAD

**Issue by kevinoid** [Azure/azure-sdk-for-net#12490]
**Describe the bug**
A class serialized with [`PolymorphicSerializeJsonConverter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.polymorphicserializejsonconverter-1) does not apply [`JsonConverterAttribute`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonConverterAttribute.htm) declared on instance properties during serialization.

**Expected behavior**
[`JsonConverterAttribute`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonConverterAttribute.htm) is applied during serialization whether or not the class is serialized using [`PolymorphicSerializeJsonConverter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.polymorphicserializejsonconverter-1).

**Actual behavior (include Exception or Stack Trace)**
[`JsonConverterAttribute`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonConverterAttribute.htm) is only applied for properties of classes which don't use [`PolymorphicSerializeJsonConverter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.polymorphicserializejsonconverter-1).

**To Reproduce**
```csharp
using System;
using Microsoft.Rest.Serialization;
using Newtonsoft.Json;

[JsonObject("animal")]
class Animal
{
    [JsonConverter(typeof(DateJsonConverter))]
    [JsonProperty("birthday")]
    public DateTime Birthday { get; set; }
}

class Program
{
    static void Main()
    {
	var animal = new Animal
	{
	    Birthday = new DateTime(2020, 2, 29),
	};

	var serializeSettings = new JsonSerializerSettings();
	serializeSettings.Converters.Add(new PolymorphicSerializeJsonConverter<Animal>("dType"));
        Console.WriteLine(JsonConvert.SerializeObject(animal, Formatting.Indented, serializeSettings));
    }
}
```
produces

    {
      "dType": "animal",
      "birthday": "2020-02-29T00:00:00"
    }

(Note that `birthday` *was not* converted using [`DateJsonConverter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.datejsonconverter).) If `serializeSettings.Converters.Add` is removed, it produces:

    {
      "birthday": "2020-02-29"
    }

(Note that `birthday` *was* converted using [`DateJsonConverter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.serialization.datejsonconverter).)

**Environment:**
 - Name and version of the Library package used: Microsoft.Rest.ClientRuntime 3.0.3 with Newtonsoft.Json 12.0.3
 - Hosting platform or OS and .NET runtime version (`dotnet --info` output for .NET Core projects): .NET Core SDK 3.1.300

**Discussion:**
The problem is that [`JsonConverterHelper.SerializeProperties`](https://github.com/Azure/azure-sdk-for-net/blob/Microsoft.Rest.ClientRuntime_2.3.21/sdk/mgmtcommon/ClientRuntime/ClientRuntime/Serialization/JsonConverterHelper.cs#L40) doesn't consider the [`JsonProperty.PropertyConverter`](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_JsonProperty_Converter.htm) when serializing.

After a bit of poking around, I don't see an easy fix without copying a lot more logic from [`JsonSerializerInternalWriter`](https://github.com/JamesNK/Newtonsoft.Json/blob/12.0.3/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs).  I also suspect there may be a lot more differences in serialization behavior due to the differences with `JsonSerializerInternalWriter.SerializeObject`, not to mention potential future divergence as features are added to Json.NET.

An alternative approach would be to use a proxy [`JsonWriter`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonWriter.htm) which doesn't pass the opening [`WriteStartObject`](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonWriter_WriteStartObject.htm) or closing [`WriteEndObject`](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonWriter_WriteEndObject.htm) and removes the "properties." prefix from names.  A potential issue with this approach is that I don't see a foolproof way to recognize additional properties (from [`JsonExtensionDataAttribute`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonExtensionDataAttribute.htm)) to avoid removing "properties.".  (I don't understand why this is done, so I'm not sure how much of an issue this is.)

Thoughts?

**Comment by kevinoid** [comment]
> I don't see an easy fix without copying a lot more logic from [`JsonSerializerInternalWriter`](https://github.com/JamesNK/Newtonsoft.Json/blob/12.0.3/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs)

Scratch that.  Using [`JsonProperty.Converter`](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_JsonProperty_Converter.htm) should be easy.  It won't fix the other potential current and future divergence from [`JsonSerializerInternalWriter`](https://github.com/JamesNK/Newtonsoft.Json/blob/12.0.3/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs), but it would fix this issue.  If that's sufficient, I can work up a PR (or I'd be happy to let someone else take it).

**Comment by jsquire** [comment]
Fixed by #12943

**Comment by jsquire** [comment]
Thanks to the kindness of @kevinoid, this comes with an associated PR (#12493).   Assigning this over to @pakrym for bookkeeping, as he's likely to be the final approval for the associated PR.

//cc: @AlexGhiondea
